drv/lpc55-spi-server/src/main.rs

```diff
@@ -63,74 +369,339 @@ fn main() -> ! {
+                || (prev_rxfull != prev_rxfull)
+                || (prev_txlvl != prev_txlvl)
+                || (prev_rxlvl != prev_rxlvl)
+            {
+                ringbuf_entry!(Trace::Loop(olc, ilc, tctx.state, rctx.state));
+                ringbuf_entry!(Trace::Irq(ssa, ssd, _mstidle));
+                ringbuf_entry!(Trace::LStat(
+                    txerr, rxerr, perint, txempty, txnotfull, rxnotempty,
+                    rxfull, txlvl, rxlvl
+                ));
+                prev_txerr = txerr;
+                prev_rxerr = rxerr;
+                prev_perint = perint;
+                prev_txempty = txempty;
+                prev_txnotfull = txnotfull;
+                prev_rxnotempty = rxnotempty;
+                prev_rxfull = rxfull;
+                prev_txlvl = txlvl;
+                prev_rxlvl = rxlvl;
+            }
+
+            // TODO: catch aborted transmissions.
+            //       state == writing and ssd
+
+            // Messages from the RoT to SP do not begin writing to FIFOWR
+            // until the SPI bus is idle (CSn not asserted).
+            // next CSn is asserted.
+            if !inframe && tctx.state == TxState::Queued {
+                tctx.state = TxState::Writing;
+                spi.txerr_clear();
+                spi.enable_tx();
+                ringbuf_entry!(Trace::StartTx(ssa, ssd));
+                gpio.set_val(rot_irq, Value::Zero).unwrap_lite();
+                tctx.rot_irq = true;
+                ringbuf_entry!(Trace::AssertRotIrq);
+            }
+
+            if tctx.state == TxState::Writing {
+                loop {
+                    if !spi.can_tx() {
+                        break;
+                    }
+                    let b = tctx.tx[tctx.count];
+                    ringbuf_entry!(Trace::Tx(tctx.state, b, tctx.count));
+                    spi.send_u8(b);
+                    tctx.count += 1;
+                    if tctx.count == tctx.end {
+                        tctx.state = TxState::Finish;
+                        break;
+                    }
+                }
+                if tctx.state == TxState::Writing {
+                    // There are more bytes to transmit
+                    again = true; // XXX needed?
                 }
-                again = true;
             }
 
-            if spi.has_byte() && !rx_done {
-                let b = spi.read_u8();
-                ringbuf_entry!(Trace::Rx(b));
-                rx[rx_cnt] = b;
-                rx_cnt += 1;
-                if rx_cnt == rx.len() {
-                    rx_done = true;
+            // TODO: check for tx underflow.
+            // TODO: check for rx overflow.
+            // TODO: Queue a failure message on those conditions.
+            // TODO: Clear the error condition.
+
+            // If we are asserting ROT_IRQ, check for conditions to de-assert.
+            if tctx.rot_irq && tctx.state == TxState::Finish {
```

Comment:
Is there any case where `tctx.state == TxState::Finish` but `tctx.rot_irq != true`?

Reply:
You are correct.